# model_evaluation exercises

## Given the following confusion matrix, evaluate (by hand) the model's performance.
|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |

### 1. In the context of this problem, what is a false positive?
    - FP = actual dog and predicted cat

### 2. In the context of this problem, what is a false negative?
    - FN = actual cat and predicted cat

### How would you describe this model?
    - This chart looks slightly different in that the predictions are the columns and the actual are the rows.

In [100]:
# solution
# cat = positive class
# dog = negative class


# False positive: We predicted cat, but it is actually a dog
# False negative: We predicted dog, but it is actually a cat

tp = 34
tn = 46
fp = 7
fn = 13

In [101]:
#solution continued

print("Cat-classifier (where 'cat' is the positive prediction)")
print("True Positives", tp)
print("False Positives", fp)
print("False Negatives", fn)
print("True Negatives", tn)

print("-------------")

accuracy = (tp + tn) / (tp + tn + fp + fn)
recall = tp / (tp + fn)
precision = tp / (tp + fp)

print("Accuracy is", accuracy)
print("Recall is", round(recall,2))
print("Precision is", round(precision,2))

Cat-classifier (where 'cat' is the positive prediction)
True Positives 34
False Positives 7
False Negatives 13
True Negatives 46
-------------
Accuracy is 0.8
Recall is 0.72
Precision is 0.83


In [1]:
import pandas as pd

In [2]:
data = {'actual_dog':['46','7'],'actual_cat':['13','34']}
df = pd.DataFrame(data, columns = ['actual_dog','actual_cat'], index=['pred_dog','pred_cat'])
df # a better way to read the data

actual_dog actual_cat
pred_dog         46         13
pred_cat          7         34

## 3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

### Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

### Use the predictions dataset and pandas to help answer the following questions:

    - An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
    
    - Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [75]:
import csv

In [76]:
df = pd.read_csv('c3.csv')

In [77]:
df.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [78]:
# They tell you that they want to identify as many of the ducks that have a defect as possible. 
# Which evaluation metric would be appropriate here? 
# Which model would be the best fit for this use case?

pd.crosstab(df.actual, df.model1) # looking at model 1

model1     Defect  No Defect
actual                      
Defect          8          8
No Defect       2        182

In [79]:
pd.crosstab(df.actual, df.model2) # looking at model 2

model2     Defect  No Defect
actual                      
Defect          9          7
No Defect      81        103

In [80]:
pd.crosstab(df.actual, df.model3) # looking at model 3

model3     Defect  No Defect
actual                      
Defect         13          3
No Defect      86         98

In [81]:
df.actual.value_counts() # looking at counts of Defects and No Defects

No Defect    184
Defect        16
Name: actual, dtype: int64

- TP = Actual "Defect" and Predicted "Defect"
- TN = Actual "No Defect" and Predicted "Defect"
- FP = Actual "Defect" and Predicted "No Defect"
- FN = Actual "No Defect" and Predicted "No Defect"

In [82]:
df['baseline_prediction'] = 'No Defect' # No defect is most frequent, adding a baseline of no defect

In [83]:
df.head() # check_yo_head

actual     model1  model2     model3 baseline_prediction
0  No Defect  No Defect  Defect  No Defect           No Defect
1  No Defect  No Defect  Defect     Defect           No Defect
2  No Defect  No Defect  Defect  No Defect           No Defect
3  No Defect     Defect  Defect     Defect           No Defect
4  No Defect  No Defect  Defect  No Defect           No Defect

In [84]:
# we want to identify the defect ducks
# use recall for model1

subset = df[df.actual == 'Defect']

model_recall = (subset.model1 == subset.actual).mean()
baseline_recall = (subset.baseline_prediction == subset.actual).mean()

print(f'   model recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model recall: 50.00%
baseline recall: 0.00%


In [85]:
# we want to identify the defect ducks
# use recall for model2

subset = df[df.actual == 'Defect']

model_recall = (subset.model2 == subset.actual).mean()
baseline_recall = (subset.baseline_prediction == subset.actual).mean()

print(f'   model recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model recall: 56.25%
baseline recall: 0.00%


In [86]:
# we want to identify the defect ducks
# use recall for model3

subset = df[df.actual == 'Defect']

model_recall = (subset.model3 == subset.actual).mean()
baseline_recall = (subset.baseline_prediction == subset.actual).mean()

print(f'   model recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model recall: 81.25%
baseline recall: 0.00%


## Which model would be the best fit for this use case?

- Model3 
- model recall: 81.25%

In [87]:
# df.replace(['No Defect', 'Defect'],  [0, 1])
# my code above was not running and I thought it was because it was not 0,1 but not needed
# make sure to check capitalization 'defect' vs 'Defect'


## They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. 

- Which evaluation metric would be appropriate here? 
- Which model would be the best fit for this use case?


In [88]:
df.head() # check_yo_head

actual     model1  model2     model3 baseline_prediction
0  No Defect  No Defect  Defect  No Defect           No Defect
1  No Defect  No Defect  Defect     Defect           No Defect
2  No Defect  No Defect  Defect  No Defect           No Defect
3  No Defect     Defect  Defect     Defect           No Defect
4  No Defect  No Defect  Defect  No Defect           No Defect

In [90]:
# testing for precision model1
subset = df[df.model1 == 'Defect']

model_precision = (subset.actual == subset.model1).mean()
print(f'model precision: {model_precision:.2%}')

model precision: 80.00%


In [91]:
subset = df[df.model2 == 'Defect']

model_precision = (subset.actual == subset.model2).mean()
print(f'model precision: {model_precision:.2%}')

model precision: 10.00%


In [92]:
subset = df[df.model3 == 'Defect']

model_precision = (subset.actual == subset.model3).mean()
print(f'model precision: {model_precision:.2%}')

model precision: 13.13%


In [93]:
model_precision

0.13131313131313133

In [94]:
# Baseline Precision: # only look at the subset where we made positive prediction
# there is no 'Defect' in the baseline_prediction column and therefore unble to do the calculation
# returns not a number 'nan'
subset = df[df.baseline_prediction == 'Defect']

baseline_precision = (subset.actual == subset.baseline_prediction).mean()
print(f'baseline precision: {baseline_precision:.2%}')

baseline precision: nan%


# 4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

## At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

### Several models have already been developed with the data, and you can find their results here.

## Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions

- a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

- b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?

- c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

In [107]:
# bringing in the new data for this question

import csv

df = pd.read_csv('gives_you_paws.csv')

In [108]:
df.head() # check_yo_head

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [109]:
df.actual.value_counts() # looking at counts of dog and cat

dog    3254
cat    1746
Name: actual, dtype: int64

In [110]:
pd.crosstab(df.actual, df.model1) # looking at model1

model1   cat   dog
actual            
cat     1423   323
dog      640  2614

In [111]:
pd.crosstab(df.actual, df.model2) # looking at model 2

model2   cat   dog
actual            
cat     1555   191
dog     1657  1597

In [112]:
pd.crosstab(df.actual, df.model3) # looking at model 3

model3   cat   dog
actual            
cat      893   853
dog     1599  1655

In [113]:
pd.crosstab(df.actual, df.model4) # looking at model 4 

model4  cat   dog
actual           
cat     603  1143
dog     144  3110

In [114]:
df['baseline_prediction'] = 'dog' # 'dog' is most frequent, adding a baseline of 'dog'

In [115]:
df.head() # check_yo_head

actual model1 model2 model3 model4 baseline_prediction
0    cat    cat    dog    cat    dog                 dog
1    dog    dog    cat    cat    dog                 dog
2    dog    cat    cat    cat    dog                 dog
3    dog    dog    dog    cat    dog                 dog
4    cat    cat    cat    dog    dog                 dog

### a. In terms of accuracy, how do the various models compare to the baseline model? 
#### Are any of the models better than the baseline?

- model1 has the highest accuracy 80.74%

In [116]:
# taking a look at model1 accuracy
model_accuracy = (df.actual == df.model1).mean()
baseline_accuracy = (df.actual == df.baseline_prediction).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 80.74%
baseline accuracy: 65.08%


In [117]:
# taking a look at model2 accuracy
model_accuracy = (df.actual == df.model2).mean()
baseline_accuracy = (df.actual == df.baseline_prediction).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 63.04%
baseline accuracy: 65.08%


In [118]:
# taking a look at model3 accuracy
model_accuracy = (df.actual == df.model3).mean()
baseline_accuracy = (df.actual == df.baseline_prediction).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 50.96%
baseline accuracy: 65.08%


In [119]:
# taking a look at model4 accuracy
model_accuracy = (df.actual == df.model4).mean()
baseline_accuracy = (df.actual == df.baseline_prediction).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 74.26%
baseline accuracy: 65.08%


In [120]:
# another solution

# get column names
models = list(df.columns)
models = models[1:]
models

# get accuracy in dictionary form for each model
output = {}
for model in models:
    accuracy = (df.actual == df[model]).mean()
    output.update({model:accuracy})
    
output

{'model1': 0.8074,
 'model2': 0.6304,
 'model3': 0.5096,
 'model4': 0.7426,
 'baseline_prediction': 0.6508}

In [121]:
pd.DataFrame(output.items(), columns = ['model', 'accuracy'] )

model  accuracy
0               model1    0.8074
1               model2    0.6304
2               model3    0.5096
3               model4    0.7426
4  baseline_prediction    0.6508

## Suppose you are working on a team that solely deals with dog pictures. 
### Which of these models would you recomend for Phase I? For Phase II?

- 'dog' = positive
- 'cat' = negative

In [122]:
df.head() #check_yo_head

actual model1 model2 model3 model4 baseline_prediction
0    cat    cat    dog    cat    dog                 dog
1    dog    dog    cat    cat    dog                 dog
2    dog    cat    cat    cat    dog                 dog
3    dog    dog    dog    cat    dog                 dog
4    cat    cat    cat    dog    dog                 dog

In [123]:
subset = df[df.actual == 'dog']
subset.head()

actual model1 model2 model3 model4 baseline_prediction
1    dog    dog    cat    cat    dog                 dog
2    dog    cat    cat    cat    dog                 dog
3    dog    dog    dog    cat    dog                 dog
5    dog    dog    dog    dog    dog                 dog
8    dog    dog    cat    dog    dog                 dog

In [124]:
# Model 1 Recall
(subset.actual == subset.model1).mean()

0.803318992009834

In [125]:
# Model 2 Recall
(subset.actual == subset.model2).mean()

0.49078057775046097

In [126]:
# Model 3 Recall
(subset.actual == subset.model3).mean()

0.5086047940995697

In [127]:
# Model 4 Recall
(subset.actual == subset.model4).mean()

0.9557467732022127

Phase I dog pictures, model 4 95.57%

In [128]:
df.head() #check_yo_head

actual model1 model2 model3 model4 baseline_prediction
0    cat    cat    dog    cat    dog                 dog
1    dog    dog    cat    cat    dog                 dog
2    dog    cat    cat    cat    dog                 dog
3    dog    dog    dog    cat    dog                 dog
4    cat    cat    cat    dog    dog                 dog

In [129]:
# creating subsets for each model to check precision


subset1 = df[df.model1 == 'dog']
subset2 = df[df.model2 == 'dog']
subset3 = df[df.model3 == 'dog']
subset4 = df[df.model4 == 'dog']

In [130]:
# Model 1 Precision
(subset1.actual == subset1.model1).mean()

0.8900238338440586

In [131]:
# Model 2 Precision
(subset2.actual == subset2.model2).mean()

0.8931767337807607

In [132]:
# Model 3 Precision
(subset3.actual == subset3.model3).mean()

0.6598883572567783

In [134]:
# Model 4 Precision
(subset4.actual == subset4.model4).mean()

0.7312485304490948

In [135]:
pd.crosstab(df.model2, df.actual) # highest percentage 89.31%

actual   cat   dog
model2            
cat     1555  1657
dog      191  1597

Phase II dog pictures: model 2 with an 89.31% precision

## Suppose you are working on a team that solely deals with cat pictures. 
### Which of these models would you recomend for Phase I? For Phase II?

- 'cat' = positive
- 'dog' = negative

In [136]:
#exercise solution

from sklearn.metrics import classification_report

In [137]:
#solution continued

x = classification_report(df.actual, df.model1,
                          labels = ['cat', 'dog'],
                          output_dict=True)
pd.DataFrame(x).T

precision    recall  f1-score    support
cat            0.689772  0.815006  0.747178  1746.0000
dog            0.890024  0.803319  0.844452  3254.0000
accuracy       0.807400  0.807400  0.807400     0.8074
macro avg      0.789898  0.809162  0.795815  5000.0000
weighted avg   0.820096  0.807400  0.810484  5000.0000

In [139]:
# model 1 solution

print("Model 1")
pd.DataFrame(classification_report(df.actual, df.model1,
                                   labels = ['cat', 'dog'],
                                   output_dict=True)).T

Model 1


precision    recall  f1-score    support
cat            0.689772  0.815006  0.747178  1746.0000
dog            0.890024  0.803319  0.844452  3254.0000
accuracy       0.807400  0.807400  0.807400     0.8074
macro avg      0.789898  0.809162  0.795815  5000.0000
weighted avg   0.820096  0.807400  0.810484  5000.0000

In [140]:
# model 2 solution


print("Model 2")
pd.DataFrame(classification_report(df.actual, df.model2,
                                   labels = ['cat', 'dog'],
                                   output_dict=True)).T

Model 2


precision    recall  f1-score    support
cat            0.484122  0.890607  0.627269  1746.0000
dog            0.893177  0.490781  0.633479  3254.0000
accuracy       0.630400  0.630400  0.630400     0.6304
macro avg      0.688649  0.690694  0.630374  5000.0000
weighted avg   0.750335  0.630400  0.631310  5000.0000

In [142]:
# model 3 solution

print("Model 3")
pd.DataFrame(classification_report(df.actual, df.model3,
                                   labels = ['cat', 'dog'],
                                   output_dict=True)).T

Model 3


precision    recall  f1-score    support
cat            0.358347  0.511455  0.421425  1746.0000
dog            0.659888  0.508605  0.574453  3254.0000
accuracy       0.509600  0.509600  0.509600     0.5096
macro avg      0.509118  0.510030  0.497939  5000.0000
weighted avg   0.554590  0.509600  0.521016  5000.0000

In [143]:
# model 4 solution

print("Model 4")
pd.DataFrame(classification_report(df.actual, df.model4,
                                   labels = ['cat', 'dog'],
                                   output_dict=True)).T

Model 4


precision    recall  f1-score    support
cat            0.807229  0.345361  0.483755  1746.0000
dog            0.731249  0.955747  0.828560  3254.0000
accuracy       0.742600  0.742600  0.742600     0.7426
macro avg      0.769239  0.650554  0.656157  5000.0000
weighted avg   0.757781  0.742600  0.708154  5000.0000

- Phase I cat pictures: Model 2 with an 89% recall

- Phase II cat pictures: Model 4 with an 81% precision